File: src/main/java/org/blitzortung/android/map/overlay/color/ColorHandler.java
package org.blitzortung.android.map.overlay.color;

import android.content.SharedPreferences;
import android.graphics.Color;
import org.blitzortung.android.app.Preferences;
import org.blitzortung.android.app.view.PreferenceKey;
import org.blitzortung.android.data.TimeIntervalWithOffset;

public abstract class ColorHandler {

	private final SharedPreferences preferences;

    private ColorScheme colorScheme;

	private ColorTarget target;

    public ColorHandler(SharedPreferences preferences) {
		this.preferences = preferences;
		updateTarget(); 
	}

	public void updateTarget() {
		target = ColorTarget.valueOf(preferences.getString(PreferenceKey.MAP_TYPE.toString(), "SATELLITE"));
        colorScheme = ColorScheme.valueOf(preferences.getString(PreferenceKey.COLOR_SCHEME.toString(), ColorScheme.BLITZORTUNG.toString()));
	}
	
	public final int[] getColors() {
		return getColors(target);
	}

	abstract protected int[] getColors(ColorTarget target);

    public int getColorSection(long now, long eventTime, TimeIntervalWithOffset timeIntervalWithOffset) {
        return getColorSection(now, eventTime, timeIntervalWithOffset.getIntervalDuration(), timeIntervalWithOffset.getIntervalOffset());
    }

    private int getColorSection(long now, long eventTime, int intervalDuration, int intervalOffset) {
        int minutesPerColor = intervalDuration / getColors().length;
		int section = (int) (now + intervalOffset * 60 * 1000 - eventTime) / 1000 / 60 / minutesPerColor;
		section = Math.min(section, getColors().length - 1);
		section = Math.max(section, 0);

		return section;
	}

    public ColorScheme getColorScheme()
    {
        return colorScheme;
    }

	public int getColor(long now, long eventTime, int intervalDuration) {
		return getColor(getColorSection(now, eventTime, intervalDuration, 0));
	}

	public int getColor(int section) {
		return getColors()[section];
	}

	public final int getTextColor() {
		return getTextColor(target);
	}
	
	public int getTextColor(ColorTarget target) {
		return 0xffffffff;
	}

    public int getLineColor() {
        switch (target) {
            case SATELLITE:
                return 0xffffffff;
            case STREETMAP:
                return 0xff000000;
        }
        return 0;
    }

    public int getBackgroundColor() {
        switch (target) {
            case SATELLITE:
                return 0xff000000;
            case STREETMAP:
                return 0xffffffff;
        }
        return 0;
    }

    public int getNumberOfColors() {
        return getColors().length;
    }

    public int[] modifyBrightness(int[] colors, float factor)
    {
        int [] result = new int[colors.length];

        float[] HSVValues = new float[3];

        for (int index=0; index<colors.length; index++) {
            Color.colorToHSV(colors[index], HSVValues);
            HSVValues[2] *= factor;
            result[index] = Color.HSVToColor(HSVValues);
        }

        return result;
    }

}

File: app/src/main/java/org/blitzortung/android/map/overlay/color/ColorHandler.kt
/*

   Copyright 2015 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.map.overlay.color

import android.content.SharedPreferences
import android.graphics.Color
import org.blitzortung.android.app.view.PreferenceKey
import org.blitzortung.android.app.view.get

abstract class ColorHandler(private val preferences: SharedPreferences) {

    var colorScheme: ColorScheme = ColorScheme.BLITZORTUNG
        private set

    private lateinit var target: ColorTarget

    init {
        updateTarget()
    }

    fun updateTarget() {
        target = ColorTarget.valueOf(preferences.get(PreferenceKey.MAP_TYPE, "SATELLITE"))
        colorScheme = ColorScheme.valueOf(preferences.get(PreferenceKey.COLOR_SCHEME, ColorScheme.BLITZORTUNG.toString()))
    }

    val colors: IntArray
        get() = getColors(target)

    protected abstract fun getColors(target: ColorTarget): IntArray

    fun getColorSection(referenceTime: Long, eventTime: Long, intervalDuration: Int): Int {
        val minutesPerColor = intervalDuration / colors.size
        var section = if (minutesPerColor > 0) ((referenceTime - eventTime) / 1000 / 60 / minutesPerColor).toInt() else 0
        return limitToValidRange(section)
    }

    fun getColor(referenceTime: Long, eventTime: Long, intervalDuration: Int): Int {
        return getColor(getColorSection(referenceTime, eventTime, intervalDuration))
    }

    fun getColor(index: Int): Int {
        return colors[limitToValidRange(index)]
    }

    private fun limitToValidRange(index: Int): Int {
        return Math.max(Math.min(index, colors.size - 1), 0)
    }

    val textColor: Int
        get() = getTextColor(target)

    open fun getTextColor(target: ColorTarget): Int {
        return when (target) {
            ColorTarget.SATELLITE -> 0xffffffff.toInt()
            ColorTarget.STREETMAP -> 0xff000000.toInt()
        }
    }

    val lineColor: Int
        get() = getLineColor(target)

    open fun getLineColor(target: ColorTarget): Int {
        return when (target) {
            ColorTarget.SATELLITE -> 0xffffffff.toInt()
            ColorTarget.STREETMAP -> 0xff000000.toInt()
        }
    }

    val backgroundColor: Int
        get() = getBackgroundColor(target)

    open fun getBackgroundColor(target: ColorTarget): Int {
        return when (target) {
            ColorTarget.SATELLITE -> 0x00000000.toInt()
            ColorTarget.STREETMAP -> 0x00ffffff.toInt()
        }
    }

    val numberOfColors: Int
        get() = colors.size

    fun modifyBrightness(colors: IntArray, factor: Float): IntArray {
        val HSVValues = FloatArray(3)

        return colors.map {
            Color.colorToHSV(it, HSVValues)
            HSVValues[2] *= factor
            Color.HSVToColor(HSVValues)
        }.toIntArray()
    }

}